perf_benchmark/benchmark_main.cc

```diff
@@ -0,0 +1,497 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+//
+#include "grpc_transcoding/request_message_translator.h"
+#include "grpc_transcoding/json_request_translator.h"
+#include "grpc_transcoding/type_helper.h"
+#include "benchmark/benchmark.h"
+#include "google/protobuf/text_format.h"
+#include "google/protobuf/struct.pb.h"
+#include "absl/strings/string_view.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/escaping.h"
+#include "absl/status/statusor.h"
+#include "absl/memory/memory.h"
+#include "google/api/service.pb.h"
+
+#include "perf_benchmark/utils.h"
+#include "perf_benchmark/benchmark_input_stream.h"
+#include "perf_benchmark/benchmark.pb.h"
+
+namespace google {
+namespace grpc {
+namespace transcoding {
+
+namespace perf_benchmark {
+namespace {
+using namespace benchmark;
+
+constexpr absl::string_view
+    kServiceConfigTextProtoFile = "benchmark_service.textproto";
+constexpr absl::string_view kNestedFieldName = "nested";
+constexpr absl::string_view kInnerMostNestedFieldName = "payload";
+constexpr absl::string_view kBytesPayloadMessageType = "BytesPayload";
+constexpr absl::string_view kStringPayloadMessageType = "StringPayload";
+constexpr absl::string_view kNestedPayloadMessageType = "NestedPayload";
+constexpr absl::string_view
+    kStructPayloadMessageType = "google.protobuf.Struct";
+constexpr absl::string_view kInt32ArrayPayloadMessageType = "Int32ArrayPayload";
+constexpr absl::string_view
+    kDoubleArrayPayloadMessageType = "DoubleArrayPayload";
+constexpr absl::string_view
+    kStringArrayPayloadMessageType = "StringArrayPayload";
+
+// Used for NestedPayload and StructPayload
+constexpr uint64_t kNumNestedLayersForStreaming = 32;
+// Used for ArrayPayload
+constexpr uint64_t kArrayPayloadLengthForStreaming = 1 << 10; // 1024
+// Used for BytesPayload
+constexpr uint64_t kBytesPayloadLengthForStreaming = 1 << 20; // 1 MiB
+// Used for Int32ArrayPayload
+constexpr uint64_t kInt32ArrayPayloadLengthForStreaming = 1 << 14; // 16384
+// Used for Segmented StringPayload
+constexpr uint64_t kSegmentedStringPayloadLengthForStreaming = 1 << 20; // 1 MiB
```

Comment:
It seems it is used for both streaming and non-streaming,  why there is streaming suffix?

Reply:
Nice catch. Segmented String is a special case where we don't have the length passed in as a variable for the non-streaming case. The num_chunks_per_msg is the variable being tested in this case.